java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java

```diff
@@ -0,0 +1,245 @@
+import org.sonar.plugins.java.api.tree.BreakStatementTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.ContinueStatementTree;
+import org.sonar.plugins.java.api.tree.DoWhileStatementTree;
+import org.sonar.plugins.java.api.tree.ForEachStatement;
+import org.sonar.plugins.java.api.tree.ForStatementTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.SwitchStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static org.sonar.plugins.java.api.tree.Tree.Kind.CONDITIONAL_AND;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.CONDITIONAL_OR;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.CONSTRUCTOR;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.IF_STATEMENT;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.METHOD;
+
+
+@Rule(key = "S3776")
+public class CognitiveComplexityMethodCheck  extends IssuableSubscriptionVisitor {
+
+  private static final int DEFAULT_MAX = 10;
+
+  @RuleProperty(
+          key = "Threshold",
+          description = "The maximum authorized complexity.",
+          defaultValue = "" + DEFAULT_MAX)
+  private int max = DEFAULT_MAX;
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(METHOD, CONSTRUCTOR);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTree method = (MethodTree) tree;
+    if (method.block() == null || ((ClassTree)method.parent()).simpleName() == null || isWithinLocalClass(method)) {
+      return;
+    }
+    CognitiveComplexityVisitor complexityVisitor = new CognitiveComplexityVisitor();
+    method.accept(complexityVisitor);
+    int total = complexityVisitor.complexity;
+    if (total > max) {
+      reportIssue(method.simpleName(), "Refactor this method to reduce its Cognitive Complexity from " + total + " to the " + max + " allowed.", complexityVisitor.flow, total - max);
+    }
+
+  }
+
+  private boolean isWithinLocalClass(MethodTree method) {
+    return hasSemantic() && method.symbol().owner().owner().isMethodSymbol();
+  }
+
+  private static class CognitiveComplexityVisitor extends BaseTreeVisitor {
+
+    private final List<JavaFileScannerContext.Location> flow;
+    private final Set<Tree> ignored;
+    int complexity;
+    int nesting;
+
+    CognitiveComplexityVisitor() {
+      complexity = 0;
+      nesting = 1;
+      flow = new ArrayList<>();
+      ignored = new HashSet<>();
+    }
+
+    private void increaseComplexity(Tree tree, int increase) {
```

Comment:
This method is currently always used to increase the complexity by the `nesting` value, or by `1`. It probably worth to have two distinguished methods to cover these 2 cases explicitly.